### ISSUE TITLE
allow SQL Server to use Azure token

### ISSUE THREAD

**Issue by chinwobble** [HangfireIO/Hangfire#1323]
When using Azure's managed service identity you need to keep fetching tokens that have a 1 hour expiry. 

Like this 
```
                var db = new SqlConnection(ConfigManager.ConnectionString);
                if (IsAzureDB)
                {
                    db.AccessToken = tokenProvider.GetAccessTokenAsync("https://database.windows.net/").GetAwaiter().GetResult();
                }
                return db;
```

In order to use this feature we need to extend `Hangfire.SqlServer.SqlServerStorage` to cater for this.
We can add an overload to the constructor that accepts `Func<SqlConnection>` which is called on `Hangfire.SqlServer.SqlServerStorage.CreateConnection()`

**Comment by gary-walker-82** [comment]
as this is specific to Azure would it not be better suited to having a Hangfire.SqlServer.AzureSqlServerStorage class?

**Comment by pieceofsummer** [comment]
This doesn't seem like a breaking change. For example, add an `Action<SqlConnection> ConnectionConfigurator` property to `SqlServerStorageOptions` and call it from `SqlServerStorage.CreateAndOpenConnection()`. We already have `ImpersonationFunc` for similar purpose in `dev` branch, so it shouldn't make too much of a difference.

@chinwobble would you like to create a pull request for this?

**Comment by chinwobble** [comment]
@pieceofsummer sure I'll create a pull request for it

**Comment by odinserj** [comment]
Yes, connection factory is great for this scenario, and I think we can release it within 1.6.X, since it can be critical on some environments.
